First cut at Raspberry Pi GPIO extensions - WIP

JordanDave opened the pull request:
Adds generalized support for GPIO extensions and specific support for the MCP23017 I2C GPIO extender.

JordanDave commented:
Not sure why the build is failing.  It's not finding the pi4j-gpio-extension files.  The jar file (pi4j-gpio-extension-1.2.jar) appears to be in the lib directory, (I think) I modified pom.xml appropriately (removed "<optional>true</optional>"), and it builds under NetBeans on my Windows machine.

pabender commented:
> Not sure why the build is failing. It's not finding the pi4j-gpio-extension files. The jar file (pi4j-gpio-extension-1.2.jar) appears to be in the lib directory, (I think) I modified pom.xml appropriately (removed "true"), and it builds under NetBeans on my Windows machine.

I see failures in multiple CI builds.  most of them are related to threading issues in tests, and may or may not be caused by the changes here (I didn't see anything obvious, but I just glanced at it quickly).

Fix the issues flagged by the static analysis first.  Some of those are just code style, but there was at least one issue I saw related to exception handling. (again, a quick glance, I didn't read through the complete list of issues).

danielb987 commented:
Static analysis failed due to:
```
java/src/jmri/jmrix/pi/RaspberryPiSensor.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
java/src/jmri/jmrix/pi/RaspberryPiTurnoutManager.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
java/src/jmri/jmrix/pi/RaspberryPiSensorManager.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
java/src/jmri/jmrix/pi/RaspberryPiTurnout.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
java/src/jmri/jmrix/pi/ProvisionMCP23017.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
java/src/jmri/jmrix/pi/RaspberryPiGpioExFactory.java:[1] (misc) NewlineAtEndOfFile: File does not end with a newline.
```
We require that java files end with a newline.

danielb987 commented:
Travis GUI failed since the files `jmri/jmrix/pi/RaspberryPiGpioExFactory.java` and `jmri/jmrix/pi/ProvisionMCP23017.java` doesn't have any tests. See this class for an example on how you can add a simple test:
https://github.com/JMRI/JMRI/blob/master/java/test/apps/AboutActionTest.java

danielb987 commented:
@icklesteve @silverailscolo @mstevetodd 
GitHub suggested you to review this PR. Can you take a look at it? Thanks.

danielb987 commented on java/src/jmri/jmrix/pi/RaspberryPiSensorManager.java:
```diff
@@ -50,15 +52,41 @@ public boolean isPullResistanceConfigurable(){
     @Override
     @Nonnull
-    public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException {
-        return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
+    public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException, IllegalArgumentException {
+        if (name.length() < 3) {
+            throw new IllegalArgumentException();
+        }
+        if (name.substring (2,3).equals (":")) {
+            return name;
+        } else {
+            return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
+        }
     }
```
Please do not throw IllegalArgumentException in this method. If the input string is bad, it is a bad system name, so BadSystemNameException should be thrown.

The code doesn't work as expected. The connection prefix might be a single letter, for example "L", or a letter followed by a number, for example "L12". So you need to use the method Manager.getSystemNamePrefix() to get the length of the system prefix and system letter:

`if (name.length() < getSystemNamePrefix().length+1) {`

---

If I understand it correctly, a system name for MCP23017 needs to be on the format:
`<system name prefix>:MCP23017:<number1>:<number2>:<number3>`

But you don't check that. The only thing you check is that the string has a colon. That's not enough. The purpose of this method is to protect the user from creating invalid system names. If an exact format is required, this method should check that the system name follows that format.

My recommendation is that you do something like this:

```
public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException {
    if ((name.length() > getSystemNamePrefix().length()+":MCP23017:".length())
            && name.substring(getSystemNamePrefix().length(), getSystemNamePrefix().length()+":MCP23017:".length()).equals(":MCP23017:")) {

        String data = name.substring(getSystemNamePrefix().length() + ":MCP23017:".length());
        String parts[] = data.split(":");
        if (parts.length != 3) throw new BadSystemNameException();

        try {
            int bus = Integer.parseInt(parts[0]);
            int channel = Integer.parseInt(parts[1]);
            int pin = Integer.parseInt(parts[2]);
            
            if (bus < 0 || bus > 255) throw new BadSystemNameException();
            if (channel < 32 || channel > 39) throw new BadSystemNameException();
            if (pin < 0 || pin > 15) throw new BadSystemNameException();
            
            return name;
        } catch (NumberFormatException ignore) {
            throw new BadSystemNameException();
        }
    }

    return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
}

```

JordanDave replied:
OK - Change to throw BadSystemNameException
OK - Use getSystemNamePrefix
No - Can't check for a three part address here.

If the MCP23017 extension was the only one possible, I could check for the three part address.  But a major goal of my code is to make these extensions user-definable.  The only requirement I want to impose (here) is that the string begin with a colon.  The user definable code (represented by ProvisionMCP23017) has code to validate the system name...for an MCP23017 device.  At the time that RaspberryPiSensorManager is running we don't know what type of device is involved.

I could add a validateSystemNameFormat entry to RaspberryPiGpioExFactory, which could invoke the validateSystemNameFormat entry for the appropriate device (ie, ProvisionMCP23017)...  I'll try making these changes and see how it comes out.

I must say that the system name handling seems arcane and fraught with less-than-ideal testing.  In this case, for example, the code suggests that the name is valid if the pin number is between 0 and 999.  Since the latest Raspberry Pi's have a total of 40 pins...and not all of them are usable...I'm not sure why we aren't checking for a much smaller range, or even figuring out what the maximum actually is.  I think I saw somewhere a reference to "9999" as well.

danielb987 replied:
> But a major goal of my code is to make these extensions user-definable.

What do you mean with that? Lets assume I (as an ordinary user without knowing how to change the source code) purchase the device `XYZ1234` and connects it to the Rasperry Pi. How does JMRI understand what `PT:XYZ1234:21:2:12` means?

Or do you mean that another developer later will add code for the device `XYZ1234`?

JordanDave replied:
JRMI doesn't understand :XYZ1234:21:2:12.  Or, for that matter, :MCP23017:1:38:6.  What JMRI does understand is that the string between the first two colons (XYZ1234 or MCP23017) should lead to a class "Provision<Type>" (ProvisionXYZ1234, ProvisionMCP23017).  That class is loaded dynamically (using class.forname).  (I haven't worked out the details, but it's my intent that the class could be added by the user, without any involvement from JMRI developers.)

In any case, everything after the second colon (21:2:12 or A,3,X) is interpreted by the appropriate Provision class and the rest of JMRI knows nothing specific about it.

Also, slightly different, but perhaps related...  The JMRI Raspberry Pi support depends on version 1.2 of Pi4j, but that's no longer being developed/supported.  Pi4j has transitioned to version 2...and version 2 doesn't include native support for GPIO extension devices.  They've also moved to a "plugin" model for extensions (though I haven't researched how their plugins work).  This may not affect JMRI as we have the full version 1.2 Pi4j but might become an issue down the road if we need to support new versions of the Raspberry Pi.

pabender replied:
> Also, slightly different, but perhaps related... The JMRI Raspberry Pi support depends on version 1.2 of Pi4j, but that's no longer being developed/supported. Pi4j has transitioned to version 2...and version 2 doesn't include native support for GPIO extension devices. They've also moved to a "plugin" model for extensions (though I haven't researched how their plugins work). This may not affect JMRI as we have the full version 1.2 Pi4j but might become an issue down the road if we need to support new versions of the Raspberry Pi.

It is probably advisable to update JMRI to use the newer version of the library now.

Relying on a library that is out of date for this work you are doing may result in more issues when the dependency is updated later.

JordanDave replied:
> It is probably advisable to update JMRI to use the newer version of the library now.

Well, yeah.  Though that's not what I had in mind when I began working on the GPIO extensions and I'm not sure I have the bandwidth to do that work in the next few months.  But I can at least take a look at the new version and maybe get some idea of the work involved in converting.

JordanDave replied:
On updating Pi4j...

First quick look.  Pi4j says they rely on JRE 11; we're running 8 (well, I'm running OpenJDK 1.8.0_212).  I added OpenJDK 11 to my RPi but ran into problems with it.  And I couldn't attach the debugger from my Windows system, so I'm not sure of the details but it appears that there are changes to reflection (which my new code is using.)

JordanDave replied:
A little more looking at Pi4j v2.  There are a few examples of plug-in code such as would be needed for MCP23017 or other GPIO extensions, but the examples are outside of the Pi4j project itself, appear to be user generated, and appear to be a long way from production quality.  If we move to Pi4j v2, and want to support extensions, it appears that we'd have to do our own development of the appropriate Pi4j plugins.

While looking in to this I also came across some of the reasoning behind their decision to remove extensions from the core project.  Their major concern had to do with long term support - folks would create an extension and add it to Pi4j but wouldn't necessarily be around to maintain it if issues arose down the road.

A problem perhaps not unfamiliar to the JMRI developer community.

JordanDave commented:
@danielb987 - Just pushed several commits could you approve the build?

Thanks

JordanDave commented:
Also... it seems as though the github builds are running under a different set of rules and restrictions from what I see in my NetBeans builds.  Is there someway I can do my own builds under the stricter rules?

danielb987 commented:
> @danielb987 - Just pushed several commits could you approve the build?

Done.

JordanDave commented:
@danielb987 - yet a couple of more minor changes.  Maybe it will build without error this time.

danielb987 commented:
> @danielb987 - yet a couple of more minor changes. Maybe it will build without error this time.

Done.

JordanDave commented:
@danielb987 - I think we're getting closer.  Maybe.  Added javadoc stuff and cleaned up some exception stuff.

Also, when I try to run the bom.xml "javadoc" target it runs for a while and then gives me the following error:

java.io.IOException: Cannot run program "C:\Program Files\Java\jdk1.8.0_271\bin\javadoc.exe": CreateProcess error=206, The filename or extension is too long

Seems to run from a command prompt...

JordanDave commented:
Well...I think my code pretty much made it through...for the first time.  There are a couple of Failures in the Windows CI Tests, but they appear to be in Logixng code.  But I know there are other issues lurking.  I'll add comments describing the issues I'm aware of.

JordanDave commented:
Concern 1 - Software outside JMRI control

- As mentioned above, JMRI ships and relies on Pi4j v 1.2.  Pi4j has migrated to v 2.x and the web site suggests that support for v 1.2 will be limited to major failures.  Whatever that means.  One of the reasons Pi4j moved to v 2.x was to remove the GPIO extensions from Pi4j itself.  Some work has been done on extensions for v 2.x, including the MCP23017, but this work appears preliminary and not under active refinement.
- As if that wasn't enough of a worry, Pi4j relies on the WiringPi library.  I believe that WiringPi ships with the Raspbian distribution, but, even so, the developer of WiringPi says he's had enough and ongoing support appears to be uncertain at best.

JordanDave commented:
Concern 2 - Save / Load Failures

I've tried saving and loading sensor, turnout and logix tables.  I had to edit out some meter and logixng items from the saved xml file, but the load seems to work.  Most of the time.  I have seen failures on occasion but have not been able to debug the failures to figure out what's going on.  It may be that the problem(s) were corrected by previous updates but I will continue to work on this if I see further failures.

If the problem(s) have not been corrected I suspect that they may have to do with questions of static vs dynamic storage and potentially with a mismatch between what our code is doing and what is going on between Pi4j and WiringPi.  I'm pretty new to Java and this is an area I don't fully understand, hence my suspicion.

JordanDave commented:
Concern 3 - Dynamically loading extension support classes

My core extension interface code is found in RaspberryPiGpioExFactory.  This code parses the pin system name to extract the string between the first two colons (ie, "MCP23017").  It uses this string to create a class name jmri.jmrix.pi.ProvisionXXX ("jmri.jmrix.pi.ProvisionMCP23017") and then loads that class using class.forName.

This approach works for ProvisionMCP23017 which is built into the JMRI Raspberry Pi code, but I don't know enough about Java class loading, etc., to know that it would work for unknown, user supplied ProvisionXXX support code.  Assuming that there's underlying Pi4j support for a device the ProvisionXXX code is the only thing a user would need to supply and there's not a lot to it.

So - I'm not sure that "jmri.jmrix.pi" is the correct, or even usable, package name.  I'm not sure where user's would put their extension code to ensure that it would be found.  And I'm not sure that the default class loader (class.forName) is the right way to load it.

These issues are pretty much ignoreable if we don't want to document the capability.

JordanDave commented:
Concern 4 - Documentation, tool tips, etc.

I'm not sure what documentation is needed, or where it goes.

While entering pin names and the like, I see different tool tips...or sometimes no tool tips...  I added a bit to the Bundle.properties in jmri.jmrix.pi but it seems that more is needed.

Also, because of the way the name verification code operates, partial name entry, the entry is flagged with an exclamation point in a red circle while it's being entered.  Once complete, the exclamation point becomes a check mark in a green circle.

danielb987 commented:
@JordanDave 
I apologize. I have been busy the last days so I haven't had time to review this PR. But I will look into this the next days.

bobjacobsen commented:
Re finding in #3:  On our three supported platforms, if you drop a .jar file into the `lib` directory, it's added to the class path at run time.  On the question of what package:  really, this is what the Service Provider architecture is meant to deal with; classes that provide services are declared in their manifest, so adding a .jar file runs their declaration code.  See the Service Provider discussion on https://www.jmri.org/help/en/html/doc/Technical/Patterns.shtml

bobjacobsen commented:
Re 4:  Documentation is always nice. Some people do a lot, some do a little.  In my opinion, getting the basics right_once_ is the most important thing. If that's done with enough tooltips to answer all questions, fine.  If that's done with a help page and no tooltips, also fine.  One complete set of info is much better than N with only 1/N of the information.  But nothing is "required" here....

bobjacobsen commented:
RE Relying on old libraries:  Depends on what kind of support is anticipated.  If it works _now_, and we're worried about the future, then we can migrate (or not) in the future.  The effort to stay at the front of the release chain for the entire world doesn't buy very much.  But if it's not getting support, and it _doesn't_ work now, and moving forward is the only way to get that resolved, then it makes sense to update ASAP.

JordanDave commented:
Concern 5 - Input pin pull resistance configuration

This one's more of a detail, but I'm not sure how to resolve it.

In general, the Raspberry Pi GPIO pins can be configured with built in pull up or pull down resistance (or none).  The MCP23017 does not have a pull down capability - it can be pull up or off.  RaspberryPiSensorManager.isPullResistanceConfigurable returns true.  I'm not sure where/how to make it clear that the MCP23017 pins only support PULL_UP and OFF ... and make sure that the limitation makes it to the UI so user's aren't confused by incorrect drop down lists, etc.

bobjacobsen commented:
Input resistance is already being handled.  For example, here's a snipped from ava/src/jmri/jmrit/beantable/sensor/SensorTableDataModel.java 
```
           case PULLUPCOL:
                PullResistanceComboBox c = new PullResistanceComboBox(Sensor.PullResistance.values());
                c.setSelectedItem(s.getPullResistance());
                return c;
```

dsand47 commented:
> Concern 2 - Save / Load Failures
> 
> I've tried saving and loading sensor, turnout and logix tables. I had to edit out some meter and logixng items from the saved xml file, but the load seems to work. Most of the time. I have seen failures on occasion but have not been able to debug the failures to figure out what's going on. It may be that the problem(s) were corrected by previous updates but I will continue to work on this if I see further failures.
> 
> If the problem(s) have not been corrected I suspect that they may have to do with questions of static vs dynamic storage and potentially with a mismatch between what our code is doing and what is going on between Pi4j and WiringPi. I'm pretty new to Java and this is an area I don't fully understand, hence my suspicion.

PanelPro xml data files are not always backward compatible.  There are two general scenarios.

1). The xml schema used for validation on an older version of JMRI does not know about changes in the xml structure of the file.  The TurnOffXmlValidation.py script in the jython directory can be used to bypass validation.
2). The xml load routines cannot handle changes to actual xml structure.

JordanDave commented:
> 
> 
> Input resistance is already being handled. For example, here's a snipped from ava/src/jmri/jmrit/beantable/sensor/SensorTableDataModel.java
> 
> ```
>            case PULLUPCOL:
>                 PullResistanceComboBox c = new PullResistanceComboBox(Sensor.PullResistance.values());
>                 c.setSelectedItem(s.getPullResistance());
>                 return c;
> ```

OK.  But, as I said, I'm relatively new to Java.

In this case what I want to do (I think) is to modify RaspberryPiSensor so that "RaspberryPiSensor.PullResistance.values()" returns 3 values for some pins (UP, DOWN and OFF) and 2 values for other pins (UP and OFF).

I don't know how to do that.

pabender commented:
> > Input resistance is already being handled. For example, here's a snipped from ava/src/jmri/jmrit/beantable/sensor/SensorTableDataModel.java
> > ```
> >            case PULLUPCOL:
> >                 PullResistanceComboBox c = new PullResistanceComboBox(Sensor.PullResistance.values());
> >                 c.setSelectedItem(s.getPullResistance());
> >                 return c;
> > ```
> 
> OK. But, as I said, I'm relatively new to Java.
> 
> In this case what I want to do (I think) is to modify RaspberryPiSensor so that "RaspberryPiSensor.PullResistance.values()" returns 3 values for some pins (UP, DOWN and OFF) and 2 values for other pins (UP and OFF).
> 
> I don't know how to do that.

One way to do that, which is used in many places in JMRI , is to create a subclass that does that.

Since this is using the pi4j library, you MAY be able to ask that library what modes are possible for the pins.  Since you are making this generic, I would see if that can work ( it has been a while since I wrote this code.  I can’t remember for certain if this is possible).

pabender commented:
> > > Input resistance is already being handled. For example, here's a snipped from ava/src/jmri/jmrit/beantable/sensor/SensorTableDataModel.java
> > > ```
> > >            case PULLUPCOL:
> > >                 PullResistanceComboBox c = new PullResistanceComboBox(Sensor.PullResistance.values());
> > >                 c.setSelectedItem(s.getPullResistance());
> > >                 return c;
> > > ```
> > 
> > 
> > OK. But, as I said, I'm relatively new to Java.
> > In this case what I want to do (I think) is to modify RaspberryPiSensor so that "RaspberryPiSensor.PullResistance.values()" returns 3 values for some pins (UP, DOWN and OFF) and 2 values for other pins (UP and OFF).
> > I don't know how to do that.
> 
> One way to do that, which is used in many places in JMRI , is to create a subclass that does that.
> 
> Since this is using the pi4j library, you MAY be able to ask that library what modes are possible for the pins. Since you are making this generic, I would see if that can work ( it has been a while since I wrote this code. I can’t remember for certain if this is possible).

It looks like Pi4j does support asking the pin object for the pull types supported
https://javadoc.io/static/com.pi4j/pi4j-core/1.2/com/pi4j/io/gpio/Pin.html

JordanDave commented:
> 
> 
> Re finding in #3: On our three supported platforms, if you drop a .jar file into the `lib` directory, it's added to the class path at run time. On the question of what package: really, this is what the Service Provider architecture is meant to deal with; classes that provide services are declared in their manifest, so adding a .jar file runs their declaration code. See the Service Provider discussion on https://www.jmri.org/help/en/html/doc/Technical/Patterns.shtml

I looked at the referenced page and, in more depth, at the Oracle "Creating Extensible Applications" tutorial.  As I'm reading it, I would structure this with an "GpioExtension" interface, which would define the capabilities that would be required of all GPIO extensions and a GpioExtensionManager which would load any available extensions and choose between them at run time.

The Extensions themselves would provide implementations for the GpioExtension interface, including some identification method (so that GpioExtensionManager could determine which extension to invoke for, say, the MCP23017 chip.)

This is functionally similar to what I've done already, but probably structured more cleanly.

So far, I'm not clear on the relationship between JMRI InstanceInitializer and java.util.ServiceLoader.

Also, would this suggest that everything would be part of the jmri.jmrix.pi package, or in a separate sub-package (jmri.jmrix.pi.extensions), or somewhere else?

JordanDave commented:
> 
> 
> @JordanDave
> I apologize. I have been busy the last days so I haven't had time to review this PR. But I will look into this the next days.

@danielb987
No problem - I'm probably more trouble than I'm worth.

Slowly beginning to know my way around some things and even maybe understand a few of them, but it's been a steep learning curve and I don't do so well with steep.

danielb987 commented:
> > @JordanDave
> > I apologize. I have been busy the last days so I haven't had time to review this PR. But I will look into this the next days.
> 
> @danielb987
> No problem - I'm probably more trouble than I'm worth.

Absolute not! I strongly appreciate the work you put into this! But I haven't had the time yet to dig into it.

JordanDave commented:
> 
> 
> > > > Input resistance is already being handled. For example, here's a snipped from ava/src/jmri/jmrit/beantable/sensor/SensorTableDataModel.java
> > > > ```
> > > >            case PULLUPCOL:
> > > >                 PullResistanceComboBox c = new PullResistanceComboBox(Sensor.PullResistance.values());
> > > >                 c.setSelectedItem(s.getPullResistance());
> > > >                 return c;
> > > > ```
> > > 
> > > 
> > > OK. But, as I said, I'm relatively new to Java.
> > > In this case what I want to do (I think) is to modify RaspberryPiSensor so that "RaspberryPiSensor.PullResistance.values()" returns 3 values for some pins (UP, DOWN and OFF) and 2 values for other pins (UP and OFF).
> > > I don't know how to do that.
> > 
> > 
> > One way to do that, which is used in many places in JMRI , is to create a subclass that does that.
> > Since this is using the pi4j library, you MAY be able to ask that library what modes are possible for the pins. Since you are making this generic, I would see if that can work ( it has been a while since I wrote this code. I can’t remember for certain if this is possible).
> 
> It looks like Pi4j does support asking the pin object for the pull types supported
> https://javadoc.io/static/com.pi4j/pi4j-core/1.2/com/pi4j/io/gpio/Pin.html

I don't have any trouble figuring out what pull ups and downs are supported...my problem is how to structure the code to make that info available to SensorTableDataModel.  The call to Sensors.PullResistance.values() is a call to an enum, which can't be subclassed...I don't think.

bobjacobsen commented:
`Sensor.PullResistance` provides the full set of valid possibilities: If there are more than UP, DOWN and OFF, we can certainly add them.

Then `setPullResistance` and `getPullResistance` do what you'd expect.

To handle a case where i.e. only UP and OFF are valid, not down, we'd add a call to Sensor (with a default implementation of `[OFF]` perhaps) that returns an array of the valid values, then subclasses could implement as needed.

JordanDave commented:
Moving on...

I've been working on a few things on a new branch - at some point I'll have to bring that back in.

In the meantime...

I've swapped the implementation of loadable GPIO extenders to use the JMRIServiceProviderInterface, @ServiceProvider, and java.util.ServiceLoader.  This seems to work as expected.

I've modified the usage of Sensors.PullResistance.values().  I added a Sensor method getAvailablePullValues with a default implementation which just returns the original "Sensors.PullResistance.values()".  I added an override to RaspberryPiSensor to use the default for regular gpio pins but to call the appropriate GPIO extender to get values for extender pins.  I also modified the RaspberryPiSensor code to change the default Pull value from "DOWN" to "OFF" - some regular GPIO pins couldn't be provisioned with the "DOWN" value.  (Pin 8, that I know of...)  When that failed, the pin was left in limbo and there was no way to create or delete it.

I (tried to?) modify the Sensor UI code to make use of getAvailablePullValues but ran into some issues here.  First, SensorPullUpEditAction creates sensorPullUpComboBox (at about line 30) initialized with Sensor.PullResistance.values().  I'm not sure how the pieces go together, but I don't think an actual sensor instance is available at that point, so I don't think the ComboBox can be instantiated with the correct values.  So I modified the "setBean" code, a few lines down, to "removeAllItems" and then add back the values from getAvailablePullValues.  This seems to work, but I'm not convinced.

In testing things, I ran into an issue with the UI on the main Sensor table window.  The table has a column for PullUp/PullDown which includes an (apparent) drop-down list.  But clicking on that drop down seems to have no effect - it just disappears into the UI code without opening the drop down list.  I'm not sure, but I don't think this is a problem with my code - can someone check to see if it happens outside of my branch?

Also in testing, I ran into a conflict between the regular Pi GPIO pins and GPIO extenders.  The Pi uses the I2C protocol to interface with the MCP23017 chip.  But the I2C protocol uses a couple of GPIO pins...and these pins are necessarily fixed - the Pi can (in theory) support a large number of separate I2C buses, operated on different GPIO pins.  The more-or-less default bus is bus #1, which uses Pi4j/WiringPi GPIO pins 8 and 9.  Creation of an MCP23017 pin (on I2C bus 1) fails if either of these pins has been used previously.  And it's not terribly obvious where the fault lies and, worse than that, it appears that deleting the offending pin(s) from JMRI doesn't do enough to make the I2C bus available.  It'll take more research to figure out to provide a better way to show the failure and how, if at all, the I2C bus can be made available again after the pin(s) have been deleted.

It's always something.

bobjacobsen commented:
> In testing things, I ran into an issue with the UI on the main Sensor table window. The table has a column for PullUp/PullDown which includes an (apparent) drop-down list. But clicking on that drop down seems to have no effect - it just disappears into the UI code without opening the drop down list. I'm not sure, but I don't think this is a problem with my code - can someone check to see if it happens outside of my branch?

I don’t have an RPi with me to test, unfortunately.  You can switch branches in git and try it with the main branch.  This is easiest when you have a clean working directory, i.e. you’ve committed everything, but even if not you can do “git stash” to put away edits and “git stash pop” to get them back.

Two things I can see in the code that might cause the problem:
 - SensorTableDataModel.isCellEditable() uses (sensorManager).isPullResistanceConfigurable() to decide if the cell can be edited, hence native JComboBox should be shown. Is that returning true always?
 - SensorTableDataModel.getPreferredWidth makes it the width of an 8-character label.  That seems narrow to me.  You could change it and see what happens.

danielb987 commented:
> In testing things, I ran into an issue with the UI on the main Sensor table window. The table has a column for PullUp/PullDown which includes an (apparent) drop-down list. But clicking on that drop down seems to have no effect - it just disappears into the UI code without opening the drop down list.

Can you see if you get an exception when this happens? And if so, give the stack trace.

JordanDave commented:
I've done some more research on the sensor table drop down issue.

The problem predates my changes.  The same issue occurs under the current production release, 4.22.

The problem is that RaspberryPiSensorManager.isPullResistanceConfigurable is never invoked.  The ProxySensorManager version is invoked instead and it always returns false.

Further research shows that this happens when the sensor table is on the "All" tab.  My setup has additional tabs for RaspberryPi, DCC++, and Internal.  If I switch to the RaspberryPi tab the Pull-Up/Down field seems to work just fine - it's

A little further testing shows that neither the DCC++ nor the Internal sensors appear to support configurable pull resistance and the drop downs on those tabs don't work either.  Also, both of them do have functioning drop downs when you use the Edit button, but changing the value in the drop down has no effect.

It seems to me that the availability of the drop down really should be by individual sensor in any case - some pins on the Raspberry Pi are more configurable than others (I don't know off hand if any are completely fixed or not, but it would not surprise me either way.)  But this would likely be a pretty major change.

Perhaps there would be a way to change the column from an (apparently) editable drop down to a display of the current value when the active sensor manager says that pull resistance is not configurable?

JordanDave commented:
@danielb987

Bunch of mostly minor commitments, etc.  Could you kick of build/tests/etc.?

Thanks

JordanDave commented:
**Concerns Revisited - Updates, etc.**

**Original Concerns**

- Concern 1 - Software outside JMRI control - Issue remains but things are working for now.
- Concern 2 - Save / Load Failures - Issue remains but does not appear to be a Raspberry Pi issue.
- Concern 3 - Dynamically loading extension support classes - Replace ad hoc approach with java ServiceLoader, etc.
- Concern 4 - Documentation, tool tips, etc. - Some work here, but issue largely remains
- Concern 5 - Input pin pull resistance configuration - Largely fixed

**New / Updated Concerns**

**Concern 6 - Conflicts in Raspberry Pi GPIO pin usage**

The Raspberry Pi supports I2C and SPI buses using standard GPIO pins.  I have not found a deterministic way to figure out exactly which pins are being used for these buses.  The default configuration of I2C bus # 1 (the one commonly used these days) uses GPIO pins 8 and 9 on my Raspberry Pi 4B.  But, as I understand it, these pins are user configurable as are pins for additional I2C buses.  Bad things happen if a user tries to use the I2C bus and also tries to provision a conflicting GPIO pin and I don't have a good way of preventing the conflict or even being sure of that a problem is the result of such a conflict.

**Concern 7 - Error Handling on Pin Provisioning Failures**

Various provisioning failures occur silently.  These need to be reported to the user, hopefully with some useful guidance.

I'm sure there are more issues that I haven't thought of...or encountered...yet.

pabender commented:
Sent from my iPhone

> On Jun 3, 2021, at 10:32 AM, JordanDave ***@***.***> wrote:
> It seems to me that the availability of the drop down really should be by individual sensor in any case - some pins on the Raspberry Pi are more configurable than others (I don't know off hand if any are completely fixed or not, but it would not surprise me either way.) But this would likely be a pretty major change.
> 

There are a lot of things that could be individually controlled by turnout or sensor ( and probably other things ) and aren’t.   

Because of the tables are Associated with individual managers, and the managers can influence the behavior of the columns, the “All” tab is a special case.  

It would be a major change to fix that .
> Perhaps there would be a way to change the column from an (apparently) editable drop down to a display of the current value when the active sensor manager says that pull resistance is not configurable?
> 

It would be better if the column was hidden if it isn’t configurable, which I thought it was...

Paul

pabender commented:
Sent from my iPhone

> On Jun 3, 2021, at 2:50 PM, JordanDave ***@***.***> wrote:
> Concern 6 - Conflicts in Raspberry Pi GPIO pin usage
> 
> The Raspberry Pi supports I2C and SPI buses using standard GPIO pins. I have not found a deterministic way to figure out exactly which pins are being used for these buses. The default configuration of I2C bus # 1 (the one commonly used these days) uses GPIO pins 8 and 9 on my Raspberry Pi 4B. But, as I understand it, these pins are user configurable as are pins for additional I2C buses. Bad things happen if a user tries to use the I2C bus and also tries to provision a conflicting GPIO pin and I don't have a good way of preventing the conflict or even being sure of that a problem is the result of such a conflict.
> 
Yesh, that is hard.  

Another hard issue is that IO pins COULD be in use for something controlled outside of JMRI ( I know some people who add a shutdown button, for example.).

Add to that that some devices have pins connected they are not using.  I seem to recall Andrew saying that about the SPROG-PI, but I could be wrong, and I can’t remember which pins....

Paul

danielb987 commented:
> Bunch of mostly minor commitments, etc. Could you kick of build/tests/etc.?

Done.

crosland commented:
>Add to that that some devices have pins connected they are not using. I 
>seem to recall Andrew saying that about the SPROG-PI, but I could be 
>wrong, and I can’t remember which pins....
Only just spotted this :)

Yes, the PI-SPROG One and 3 (but not 3v2 or any others) have connections 
from SPI MOSI (GPIO pin 19), MISO(21), SCLK(23) and CE0(24).

BUT, they have never been used. MOSI, SCLK and CE0 are PIC inputs only, 
never driven and ignored. MISO is connected to a PIC output but it is 
left tri-state (high impedance).

I have no plans to ever use them, so it is safe to assume these pins are 
available.

PI-SPROG One and 3 connects to RXD(8) and TXD(10), my other hardware 
also connects to CTS(36) and RTS(11). These pins are not available for 
other uses.

Andrew

JordanDave commented:
@danielb987

Three more commits to (hopefully) satisfy build checks.  Thanks...

danielb987 commented:
> Three more commits to (hopefully) satisfy build checks. Thanks...

Done.

danielb987 commented:
Spotbugs failed due to:
```
1. ERROR in /home/runner/work/JMRI/JMRI/java/src/jmri/jmrix/pi/extendgpio/ExtensionService.java (at line 27)
	private static ExtensionService service;
                                        ^^^^^^^
The value of the field ExtensionService.service is not used
```

JordanDave commented:
@danielb987
Fixed the unused variable.

What do I need to do to catch the (wide variety of) errors that show up here?  They don't show up when I do a regular build from within Netbeans.  I do see a variety of "Project Configurations", including test-warning-check and travis-spotbugs - which of the configurations should I check, or is there one which covers the same ground as the builds here?

danielb987 commented:
Build is restarted.

I'm not sure which test you should run. travis-spotbugs seems to run spotbugs. I suggest you try them. I usually create a PR and let the CI tests find the bugs for me, at least spotbugs errors.

bobjacobsen commented:
```
ant warnings
ant spotbugs
ant tests-warnings
```

JordanDave commented:
@danielb987
One more time...

danielb987 commented:
> One more time...

No problem  :)
Done.

JordanDave commented:
@danielb987
I think I fixed the javadoc issues from the latest build.

I don't think the Windows CI errors are mine...

`Error:  Failures: 
Error:    LayoutEditorToolsTest.testGetHeadFromNameEmptyName Multiple Failures (2 failures)
	jmri.NamedBean$DuplicateSystemNameException: systemName is already registered: IH0
	java.lang.NullPointerException: <no message>`

danielb987 commented:
> I don't think the Windows CI errors are mine...

We have many failures if Windows CI so I agree on this. The same applies to Travis there we too have a lot of failures.

JordanDave commented:
The latest commits seem to have built and made it through the Windows CI tests.  Yay.  But it failed in the travis-ci build.

This failure is a result of GpioExtensionTest, which doesn't actually cover any of the (very limited) code in GpioExtension.java.  I'm not sure how to test GpioExtension as it's an interface not a class.  It does have one default method which I might be able to test, but I'm not sure how to actual invoke that method.  Maybe I need to create an implmentation within GpioExtensionTest.java?

JordanDave commented:
@danielb987
I added stuff to GpioExtensionTest so it actually invokes a bit of GpioExtension, so travis-ci seems happy.  Please kick off a build.

Still needs documentation and better error handling.

danielb987 commented:
> I added stuff to GpioExtensionTest so it actually invokes a bit of GpioExtension, so travis-ci seems happy. Please kick off a build.

Done.

JordanDave commented:
FYI -

With things apparently building and running here I decided to try to put together another GPIO extension.  I had an MCP23008 chip lying around - basically the same as the 23017 but only 8 GPIO pins rather than 16.  I worked completely outside of my normal JMRI development environment, just using "javac" and "jar" on my Raspberry Pi.  It took four or five hours to figure out the ins and outs of the process, but I ended up with a jar file with MCP23008GpioExtension, it's nested classes, and a META-INF/services/jmri.jmrix.pi.extendgpio.spi.GpioExtension file.  I copied the resulting jar file into the JMRI lib directory, ran PanelPro normally, and had access to the 23008 chip.  (To make sure, I created a turnout and hooked up an LED - changing the turnout state changes the LED state.

So that part of my plan seems to have actually worked.  No one is more surprised than I am.

JordanDave commented:
I've taken a first cut at documenting all of this - from use of the MCP23017 to user creation and installation of MCP23008 support.

Here's the doc - just in text format for the moment:
[RaspberryPiExtensionsDoc.txt](https://github.com/JMRI/JMRI/files/6609379/RaspberryPiExtensionsDoc.txt)

The doc references two files.  The first is the GpioExtension.java interface definition:
[GpioExtension.txt](https://github.com/JMRI/JMRI/files/6609449/GpioExtension.txt)

The second is the MCP23008GpioExtension.java source file:
[MCP23008GpioExtension.txt](https://github.com/JMRI/JMRI/files/6609456/MCP23008GpioExtension.txt)

bobjacobsen commented:
You can link to a source file with a link like this:

https://github.com/JMRI/JMRI/blob/master/java/src/jmri/Meter.java

But it’s often better to link to the Javadoc:

https://www.jmri.org/JavaDoc/doc/jmri/Meter.html

Those will then populate automatically when your code and documentation are merged.

JordanDave commented:
**Raspberry Pi System Name Input Handling**

The current implementation for turnout and sensor system name input is sometimes less than helpful, especially with the addition of extended GPIO addresses.  Tool tips seem to come and go, the little red, amber and green icons in the address input box don't
have obvious meanings, and error messages tend to be generic without providing details to help the user resolve any problems.

I'm not sure how changes might be implemented but here's a general idea of what I think might be helpful as the user enters an extended GPIO address.

- User enters initial colon.  Icon goes to green checkmark, tool tip goes to something like "enter extension type followed by a colon"
  
- User continues to enter text.  Icon and tool tip remain as above.  (Currently icon goes to red exclamation point and tool tip goes to "null"

- User enters second colon.  JMRI verifies existance of the named extension.  If it exists icon remains as green checkmark, tool tip goes to something specific to the extension.  (for the MCP23017 that might be something like "enter I2C bus number, I2C channel number, and device pin number, separated by colons")  If the named extension does not exist, pop up an error dialog, change the icon to red exclamation point and change the tool tip to something like "named extension not found".
  
- Once the extension name has been entered and the extension has been found, query the extension as additional characters are entered to verify if the partially entered data is plausible.  If so, continue normal data entry.  If not, display an extension
  specific error dialog.  Plausibility might vary depending on what the extension can say about a partially entered address.  For the MCP23017, it might not be able to tell if a specific I2C bus number is valid, but it might be able to say that the bus number
  is limited to a single digit.  Likewise, without knowing if the device is located on a specific channel, it might be able to flag an out of range channel if it's not between 32 and 39 (the range supported by the MCP23017).
  
- Problems may still show up after the user has entered the complete name and selected the "Create" button.  The I2C channel might be plausible, but not configured.  The I2C channel number might be usable, but the device might not be configured on that channel.  These sorts of errors should result in expansion device specific error dialogs and the icon and tool tip should be updated appropriately.
  
As I said, I'm not sure what would need to be done to implement such an approach - I get kind of lost in all the GUI layers - but the current situation isn't viable - entering an MCP23017 address which references a non-existant I2C bus just sits there with the red exclamation point and the "null" tool tip...and doesn't allow the "Create" button to be selected, so the user has no further options.  They know something's wrong, but not what that might be.

JordanDave commented:
@danielb987
@bobjacobsen
@pabender

The recent discussion about moving to Java 11 may make much of this work more-or-less obsolete, or at least, may require significant additional work.  The current version (v1.2) of Pi4j relies on Java 1.8 and uses sun.misc.SharedSecrets somewhere in the process of supporting GPIO extensions like MCP23017.  It breaks when run under Java 11.  In addition, the Pi4j project itself appears to be in the midst of a significant transition - they both recommend for and against the use of their latest version v2.0.  (They suggest new projects use it but also suggest that it not be used in a production environment.)  v2.0 runs under Java 11 but doesn't include support for GPIO extension devices like the MCP chips.

They have an intermediate version, v1.4, which also runs under Java 11 but, like 2.0, they've stripped extension support out from Pi4j itself.

They advertise the ability to add plugin device support to both v1.4 and v2.0, but I haven't seen any clear documentation of how this would be accomplished.

Finally, v2.0 of Pi4j moves from the WiringPi hardware interface to the PiGPIO hardware interface which appears to result in a complete change in the GPIO pin numbering scheme.  (One of the Pi's most annoying traits has been the confusion over GPIO pin numbering.  Ah well.)  There may be a way around the change - perhaps some sort of post-processing on our saved data files could tweak the numbers between the old and new schemes.

In any case, I'm coming to the conclusion that further work on extension support should be delayed until the Java and Pi4j version issues are worked out.

icklesteve commented:
> **Raspberry Pi System Name Input Handling**
> 
> The current implementation for turnout and sensor system name input is sometimes less than helpful, especially with the addition of extended GPIO addresses. Tool tips seem to come and go, the little red, amber and green icons in the address input box don't
> have obvious meanings, and error messages tend to be generic without providing details to help the user resolve any problems.

Hi @JordanDave

Improving input address validation within the BeanTables is a long standing issue, see  #2664  #7661 
and clearly outside of scope for this PR.

Any of the methods which throw IllegalArgumentException can be improved by instead throwing a https://www.jmri.org/JavaDoc/doc/jmri/NamedBean.BadSystemNameException.html which can contain text like "Channel 2 out of range for MCP23017"

I'd also be tempted to add unit tests for valid addresses ( to make sure they stay valid when there is improved validation. )

Fingers crossed you get the Pi4j issue sorted,

Steve.
